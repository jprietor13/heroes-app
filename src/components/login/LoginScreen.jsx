import React, { useContext } from 'react'
import { types } from '../../types/types';
import { AuthContext } from '../../auth/authContext';
import { useForm } from '../../hooks/useForm';
import { useNavigate } from 'react-router-dom'

export const LoginScreen = () => {

  const navigate = useNavigate();
  const { dispatch } = useContext(AuthContext)
  const [ formData, handleInputChange ] = useForm({
    userName: ''
  })
  const { userName } = formData

  const handleLogin = () => {
    if(userName.trim().length < 1) {
      return null
    }

    const data = {
      name: userName,
    }

    const action = {
      type: types.login,
      payload: data
    }

    dispatch(action)

    navigate('/', { replace: true })

  }

  return (
    <div className='container mt-5'>
      <h1>Login</h1>
      <hr />
      <input 
        type="text" 
        className='form-control'
        placeholder='Nombre Usuario'
        autoComplete='off'
        name="userName"
        value={userName}
        onChange={handleInputChange}
      />
      <button type="button" className='btn btn-primary mt-2' onClick={handleLogin}>Login</button>
    </div>
  )
}
