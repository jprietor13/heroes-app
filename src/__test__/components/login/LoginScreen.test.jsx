import { mount } from "enzyme";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import { AuthContext } from "../../../auth/authContext";
import { LoginScreen } from "../../../components/login/LoginScreen";
import { types } from "../../../types/types";

const mockNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate, 
}))


describe('Pruebas en <Navbar />', () => {

  const contextValue = {
    dispatch: jest.fn(),
    user: {
      logged: false
    }
  }

  const wrapper = mount(
    <AuthContext.Provider value={contextValue}>
      <MemoryRouter initialEntries={ ['/login'] }>
        <Routes>
          <Route path="/login" element={ <LoginScreen /> }/>
        </Routes>
      </MemoryRouter>
    </AuthContext.Provider>
  )

  it('Debe mostrarse correctamente', () => {

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('h1').text().trim()).toBe('Login')

  })

  it("Debe de realizar el dispatch y la navegacion", () => {

    //obteniendo el valor del input
    wrapper.find('input').simulate('change', {
      target: {
        name: 'userName',
        value: 'Juan Prieto'
      }
    })

    const handleClick = wrapper.find('button').prop('onClick')
    handleClick()
   
    expect(contextValue.dispatch).toHaveBeenCalledWith({
      type: types.login,
      payload: {
        name: "Juan Prieto"
      }
    })

    expect(mockNavigate).toHaveBeenCalledWith( "/marvel", {replace: true} )
    localStorage.setItem('lastPath', '/dc')

    handleClick()
    expect(mockNavigate).toHaveBeenCalledWith( "/dc", {replace: true} )
    
  })

})
