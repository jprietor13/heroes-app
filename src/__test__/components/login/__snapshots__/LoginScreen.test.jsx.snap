// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Navbar /> Debe mostrarse correctamente 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login
  </h1>
  <hr />
  <input
    autoComplete="off"
    className="form-control"
    name="userName"
    onChange={[Function]}
    placeholder="Nombre Usuario"
    type="text"
    value=""
  />
  <button
    className="btn btn-primary mt-2"
    onClick={[Function]}
    type="button"
  >
    Login
  </button>
</div>
`;
